script: Fix panic in `htmlimageelement.rs` using `str::find()` to find character boundaries.

NotnaKO opened the pull request:
In function collect_sequence_characters replace loop with .chars().enumerate() by .find and split with its index.
In function parse_a_srcset_attribute many fixes in algo https://html.spec.whatwg.org/multipage/images.html#parsing-a-srcset-attribute:

1) replace checking whitespace to checking ascii_whitespaces according to the algo
2) check more carefully if in 8
3) replace buffered variable to peekable iterator
4) drop chars().enumerate() in descriptor's loop in 8
5) rewrite loop 13: fix parsing digits and remaining(w|x|h), add checks in guards that whole numbers are parsed (because functions parse_unsigned_integer and parse_double parse only valid prefix of input)

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #32976 

<!-- Either: -->
- [X] There are tests for these changes OR

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Commit:
fix loop with chars().enumerate() by using find()

Signed-off-by: Kopanov Anton <anton.kopanov@ya.ru>

mrobinson commented:
This has the same failures that showed up in #32978, so maybe we can close this and discuss fixing them there?

NotnaKO commented:
@mrobinson I think now it fix parse_a_srcset_attribute

mrobinson commented:
@NotnaKO Thanks. Can you describe the changes you made in the PR body as well as the adaptations you had to make in `parse_a_srcset_attribute`? This will help make it easier to review your change and will make it more useful for people looking at the commit history.

NotnaKO commented:
> @NotnaKO Thanks. Can you describe the changes you made in the PR body as well as the adaptations you had to make in `parse_a_srcset_attribute`? This will help make it easier to review your change and will make it more useful for people looking at the commit history.

Now better?

mrobinson commented:
> > @NotnaKO Thanks. Can you describe the changes you made in the PR body as well as the adaptations you had to make in `parse_a_srcset_attribute`? This will help make it easier to review your change and will make it more useful for people looking at the commit history.
> 
> Now better?

Yes, thanks. This is helpful.

NotnaKO commented:
@mrobinson Can you add the label to the tests again, please? (I didn't know about ./mach fmt, but I fixed it now, sorry for that)

Taym95 commented:
> @mrobinson Can you add the label to the tests again, please? (I didn't know about ./mach fmt, but I fixed it now, sorry for that)

easy way if to run this label/workflow on your fork

NotnaKO commented:
> Test results for linux-wpt-layout-2020 from try job ([#10350103714](https://github.com/servo/servo/actions/runs/10350103714)):
> 
> Flaky unexpected result (19)
> Stable unexpected results that are known to be intermittent (12)
> Stable unexpected results (7)
> * OK `/html/semantics/embedded-content/the-img-element/current-pixel-density/basic.html`* FAIL [expected PASS] subtest: `&lt;img srcset="/images/green-256x256.png 9e99999999999999999999999x" data-expect="0"&gt;`
>     ```
>     assert_equals: width expected 0 but got 28
>     ```
> * PASS [expected FAIL] `/html/semantics/embedded-content/the-img-element/sizes/sizes-auto-rendering-2.html`
> * FAIL [expected PASS] `/html/semantics/embedded-content/the-img-element/sizes/sizes-auto-rendering-3.html`
> * FAIL [expected PASS] `/html/semantics/embedded-content/the-img-element/sizes/sizes-auto-rendering.html`
> * OK [expected CRASH] `/html/semantics/embedded-content/the-img-element/srcset/parse-a-srcset-attribute.html`
> * OK `/html/semantics/embedded-content/the-img-element/update-the-image-data/fail-to-resolve.html`* PASS [expected FAIL] subtest: `&lt;img srcset="//[, /images/red.png"&gt;`
> * OK `/html/semantics/embedded-content/the-img-element/update-the-source-set.html`* PASS [expected FAIL] subtest: `&lt;img src="data:,a" srcset="data:,b 1e0x" data-expect="data:,b"&gt;`

I didn't quite understand, those tests that fell, although it was expected to pass, is this a problem or are the tests like this? Auto-rendering tests write that 429 pixels have changed, how can I understand something is wrong or not? And another test basic.html, which from 9e9999999999999999999999x probably got 0, because there was a number parsing error in the previous version of the algorithm (it seems that 9 was taken there, then there was e, not x, which made an error).

mrobinson commented:
> I didn't quite understand, those tests that fell, although it was expected to pass, is this a problem or are the tests like this? Auto-rendering tests write that 429 pixels have changed, how can I understand something is wrong or not? And another test basic.html, which from 9e9999999999999999999999x probably got 0, because there was a number parsing error in the previous version of the algorithm (it seems that 9 was taken there, then there was e, not x, which made an error).

A test failure typically indicates a bug in the implementation. The first step is to understand what part of the specification the test is trying to test and the second step is to figure out why our implementation doesn't match up with what the specification says. In very rare cases the expected behavior does not match up with what the specification says, but this is very rare.

NotnaKO commented:
Ok, I check the test auto-rendering.html and now I more sure that old version is wrong and probably test too. In test we have:
```js
testImg.setAttribute('srcset', `
    /images/fail.gif 10w,
    /images/green.png 100w,
    /images/red.png 1000w
  `);
```
So we have a call of the function `parse_a_srcset_attribute` with input 
```rs
"
                /images/fail.gif 10w,
                /images/green.png 100w,
                /images/red.png 1000w
"
```
, right?
I checked descriptors in the old version and they are 
```
["10w"]
["100w"]
["1000w\n"]
```
Here you can see the last \n. It is here, because in old version we check just space (although in html standard we have ascii_whitespace, which is not just space). In other descriptors we don't have \n, because we have a comma after them. So the remaining is not equal  "w" and we have an error in last descriptor.
That's why result of the old version of function is  
```rs
[ImageSource { url: "/images/fail.gif", descriptor: Descriptor { wid: Some(10), den: None } },
 ImageSource { url: "/images/green.png", descriptor: Descriptor { wid: Some(100), den: None } }]
```
And new version returns 
```rs
[ImageSource { url: "/images/fail.gif", descriptor: Descriptor { wid: Some(10), den: None } }, 
ImageSource { url: "/images/green.png", descriptor: Descriptor { wid: Some(100), den: None } }, 
ImageSource { url: "/images/red.png", descriptor: Descriptor { wid: Some(1000), den: None } }]
```
That's why I have different 429 pixels in test.

jdm commented:
Oh, to be clear: when it says "expected FAIL", it means that Servo's current stable test result is a failure (for the reasons that you described). In that case, a PASS expected FAIL result is a good outcome, and means we can update the stable results to reflect that Servo us now expected to pass the test. https://book.servo.org/hacking/testing.html#updating-web-test-expectations

jdm commented:
But yeah, a new FAIL expected PASS result is surprising. I misread which tests you were investigating.

NotnaKO commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -1891,164 +1891,143 @@ pub fn collect_sequence_characters<F>(s: &str, predicate: F) -> (&str, &str)
-            }
-            match state {
-                ParseState::InDescriptor => match next_char {
-                    Some((_, ' ')) => {
-                        if !current_descriptor.is_empty() {
-                            descriptors.push(current_descriptor.clone());
-                            current_descriptor = String::new();
-                            state = ParseState::AfterDescriptor;
-                        }
-                        continue;
```
I think this is the bug with '/n'. We need char::is_ascii_whitespace check like in new version.

NotnaKO commented:
I also discover test basic.html, and find a very strange behaivor in parse_double function:
I see I call parse_double(9e99999999999999999999999) and result is Ok(9.0). This is a bug in function read_exponent (in style::str) in parse_double in line:
```rs
read_numbers(iter).0.map(|exp| exp.to_i32().unwrap_or(0))
```
We read 99999999999999999999999 and it doesn't convert to i32, so return 0 and we have Ok(9.0).

mrobinson committed:
Add documentation to parser and fix some small issues

- Rename the properties of `Descriptor` so that they are full words
- Use the Rust-parser to parse doubles
- Add documentation and restructure parser to be more readable

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mrobinson commented:
I have pushed a new commit to this branch that:

1. Reworks a bit of the algorithm that you have written and added specification language to make it more readable. I needed to do this to understand that the code was following the specification and what it was doing.
2. Switched the double parsing to use Rust's builtin parser. The HTML specification has different requirements for parsing than the CSS specification. I've added a comment giving more details about the situation -- essentially we need to filter strings before we parse them like Gecko does in https://searchfox.org/mozilla-central/rev/a85b25946f7f8eebf466bd7ad821b82b68a9231f/dom/base/ResponsiveImageSelector.cpp#36.
3. Updated test expectations. The two new failures are due to the fact that we do not support the `auto` value for the `sizes` attribute. Being better at parsing `srcset` uncovered these failures.

mrobinson commented:
It looks like one of the pre-existing unit tests failed because it expected an invalid `srcset` string to be parsed. I've updated the test and pushed to the branch again.

jdm commented:
```
 ▶ OK /html/semantics/embedded-content/the-img-element/srcset/select-an-image-source.html
  │   ▶ FAIL [expected PASS] "data:,a 1w 1h, data:,b 1w"
  │   │   → assert_equals: expected "data:,a" but got "data:,b"
  │   │ FAIL [expected PASS] "data:,a 1w 1h, data:,b 1w 2h"
  │   │   → assert_equals: expected "data:,a" but got ""
  │   │ FAIL [expected PASS] "data:,a 1w 2h, data:,b 1w 1h"
  │   │   → assert_equals: expected "data:,a" but got ""
  │   │ 
  │   │ check/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:18:18
  │   │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2622:25
  │   │ test@http://web-platform.test:8000/resources/testharness.js:633:30
  │   │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:11:7
  └   └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:23:14
```

mrobinson commented:
> ```
> 
>  ▶ OK /html/semantics/embedded-content/the-img-element/srcset/select-an-image-source.html
> 
>   │   ▶ FAIL [expected PASS] "data:,a 1w 1h, data:,b 1w"
> 
>   │   │   → assert_equals: expected "data:,a" but got "data:,b"
> 
>   │   │ FAIL [expected PASS] "data:,a 1w 1h, data:,b 1w 2h"
> 
>   │   │   → assert_equals: expected "data:,a" but got ""
> 
>   │   │ FAIL [expected PASS] "data:,a 1w 2h, data:,b 1w 1h"
> 
>   │   │   → assert_equals: expected "data:,a" but got ""
> 
>   │   │ 
> 
>   │   │ check/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:18:18
> 
>   │   │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2622:25
> 
>   │   │ test@http://web-platform.test:8000/resources/testharness.js:633:30
> 
>   │   │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:11:7
> 
>   └   └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/srcset/common.js:23:14
> 
> ```

Hrm. Looks like there's a bit more work to do here.
